packages/web3-common/src/formatter.ts

```diff
@@ -0,0 +1,253 @@
+	format: DataFormat,
+) => {
+	// If it's a scalar value
+	if (!isObject(data) && !Array.isArray(data)) {
+		return convertScalarValue(data, schema?.eth as string, format);
+	}
+
+	const object = data as Record<string, unknown>;
+
+	for (const [key, value] of Object.entries(object)) {
+		dataPath.push(key);
+		const schemaProp = findSchemaByDataPath(schema, dataPath);
+
+		// If value is a scaler value
+		if (schemaProp === undefined) {
+			delete object[key];
+			dataPath.pop();
+
+			continue;
+		}
+
+		// If value is an object, recurse into it
+		if (isObject(value)) {
+			// dataPath.push(key);
+			convert(value, schema, dataPath, format);
+			dataPath.pop();
+			continue;
+		}
+
+		// If value is an array
+		if (Array.isArray(value)) {
+			if (schemaProp?.items === undefined) {
+				// Can not find schema for array item, delete that item
+				delete object[key];
+				dataPath.pop();
+
+				continue;
+			}
+
+			// If schema for array items is a single type
+			if (isObject(schemaProp.items) && schemaProp.items.eth !== undefined) {
+				for (let i = 0; i < value.length; i += 1) {
+					(object[key] as unknown[])[i] = convertScalarValue(
+						value[i],
+						// eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+						schemaProp?.items?.eth as string,
+						format,
+					);
+				}
+
+				dataPath.pop();
+				continue;
+			}
+
+			// If schema for array items is an object
+			if (
+				!Array.isArray(schemaProp?.items) &&
+				(schemaProp?.items as JsonSchema).type === 'object'
+			) {
+				for (const arrObject of value) {
+					convert(
+						arrObject as Record<string, unknown> | unknown[],
+						schema,
+						dataPath,
+						format,
+					);
+				}
+
+				dataPath.pop();
+				continue;
+			}
+
+			// If schema for array is a tuple
+			if (Array.isArray(schemaProp?.items)) {
+				for (let i = 0; i < value.length; i += 1) {
+					(object[key] as unknown[])[i] = convertScalarValue(
+						value[i],
+						(schemaProp.items as JsonSchema[])[i].eth as string,
+						format,
+					);
+				}
+
+				dataPath.pop();
+				continue;
+			}
+		}
+
+		object[key] = convertScalarValue(value, schemaProp.eth as string, format);
+
+		dataPath.pop();
+	}
+
+	return object;
+};
+
+export const format = <
+	DataType extends Record<string, unknown> | unknown[] | unknown,
+	ReturnType extends DataFormat,
+>(
+	schema: ValidationSchemaInput | JsonSchema,
+	data: DataType,
+	returnFormat: ReturnType,
+): FormatType<DataType, ReturnType> => {
+	let dataToParse: Record<string, unknown> | unknown[] | unknown;
+
+	if (isObject(data)) {
+		dataToParse = mergeDeep({}, data);
+	} else if (Array.isArray(data)) {
+		dataToParse = [...data];
+	} else {
+		dataToParse = data;
+	}
+
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+	const jsonSchema: JsonSchema = isObject(schema) ? schema : utils.ethAbiToJsonSchema(schema);
+
+	if (!jsonSchema.properties && !jsonSchema.items && !jsonSchema.eth) {
+		throw new Error('Nazar');
```

Comment:
Nazar, you're a great developer, not an error :slightly_smiling_face:

Reply:
There is a lot of other refactoring coming through, will refactor these error constructors as well.